eventdemo/source/c/src/kaa_demo.c

```diff
@@ -19,212 +19,155 @@
-            int32_t *degree = (int32_t *) info->degree->data;
-            printf("Degree=%d\n", *degree);
-        }
-        if (info->target_degree->type == KAA_THERMOSTAT_EVENT_CLASS_FAMILY_UNION_INT_OR_NULL_BRANCH_0) {
-            int32_t *target_degree = (int32_t *) info->target_degree->data;
-            printf("Target degree=%d\n", *target_degree);
-        }
-        if (info->is_set_manually->type == KAA_THERMOSTAT_EVENT_CLASS_FAMILY_UNION_BOOLEAN_OR_NULL_BRANCH_0) {
-            int8_t *is_set_manually = (int8_t *) info->is_set_manually->data;
-            printf("Is_set_manually=%s\n", (*is_set_manually) ? "true" : "false");
-        }
+    pthread_mutex_lock(&lock);
+    room_t *t = malloc(sizeof(room_t));
+    if (t == NULL) {
+        printf("Out of memory\n");
+        exit(EXIT_FAILURE);
     }
-
-    event->destroy(event);
+    strncpy(t->room_name, room_name, strlen(room_name) + 1);
+    t->next = NULL;
+    if (lst->head == NULL)
+    {
+        lst->head = t;
+        lst->tail = t;
+        pthread_mutex_unlock(&lock);
+        return;
+    }
+    lst->tail->next = t;
+    lst->tail = t;
+    pthread_mutex_unlock(&lock);
 }
 
-void kaa_on_change_degree_request(void *context
-                                , kaa_thermostat_event_class_family_change_degree_request_t *event
-                                , kaa_endpoint_id_p source)
+void pop(list_t *lst, const char *room_name)
 {
-    KAA_DEMO_UNUSED(context);
-    KAA_DEMO_UNUSED(source);
-
-    printf("ChangeDegreeRequest event received!\n");
-
-    if (event->degree->type == KAA_THERMOSTAT_EVENT_CLASS_FAMILY_UNION_INT_OR_NULL_BRANCH_0) {
-        int32_t *degree = (int32_t *) event->degree->data;
-        printf("Change temperature by %d degrees\n", *degree);
+    pthread_mutex_lock(&lock);
+    room_t *tmp, *prev_tmp = NULL;
+    for (tmp = lst->head; tmp != NULL; prev_tmp = tmp, tmp = tmp->next) {
+        if (!strncmp(room_name, tmp->room_name, strlen(room_name))) {
+            if (tmp == lst->head && tmp == lst->tail) {
+                initialize_List(lst);
+                free(tmp);
+                goto exit;
+            }
+            if (prev_tmp != NULL && tmp != lst->tail) {
+                prev_tmp->next = tmp->next;
+                free(tmp);
+                goto exit;
+            }
+            if (prev_tmp == NULL && tmp != lst->tail) {
+                lst->head = tmp->next;
+                free(tmp);
+                goto exit;
+            }
+            if (tmp == lst->tail) {
+                lst->tail = prev_tmp;
+                prev_tmp->next = NULL;
+                free(tmp);
+                goto exit;
+            }
+        }
     }
-
-    event->destroy(event);
+exit:
+    pthread_mutex_unlock(&lock);
+    return;
 }
 
-
-
-/*
- * Callback-s which receive result of event listener requests.
- */
+void print_list(const list_t *lst)
+{
+    printf("Available rooms:\n");
```

Comment:
List printing during push()/pop() operation by another thread can lead to UB here. Thus, requires mutex locking as well.